compiler/rustc_middle/src/mir/syntax.rs

```diff
@@ -1480,13 +1480,15 @@ pub enum BinOp {
     BitOr,
     /// The `<<` operator (shift left)
     ///
-    /// The offset is truncated to the size of the first operand and made unsigned before shifting.
+    /// The offset (RHS) is re-interpreted as an unsigned value of the same size and then truncated
+    /// to the size of the first operand before shifting.
```

Comment:
I gave it another try, this time working entirely on mathematical integers.

Reply:
> (Or we just take the Shl out entirely and lower << to BitAnd+ShlUnchecked :P)

It would have to involve a cast as well.

The latest variant of spelling this out actually doesn't seem so bad, it'll look nice even in MiniRust (where we work on mathematical integers). :D

Reply:
You need the cast because otherwise the `BitAnd` will not be enough I think?
Ah I guess currently that can't happen as we don't have types that are big enough. But once we have 256-bit-size types, a BitAnd on an i8 will not make it positive so BitAnd + ShlUnchecked can have UB. `Shl` as specified will keep working correctly for arbitrarily large types.